src/Control/Effect/NonDet.hs

```diff
@@ -12,81 +12,121 @@ module Control.Effect.NonDet
+runNonDet (NonDetC m) = m (liftA2 (<|>)) (pure . pure) (pure empty)
 
 -- | A carrier for 'NonDet' effects based on Ralf Hinze’s design described in [Deriving Backtracking Monad Transformers](https://www.cs.ox.ac.uk/ralf.hinze/publications/#P12).
 newtype NonDetC m a = NonDetC
-  { -- | A higher-order function receiving two parameters: a function to combine each solution with the rest of the solutions, and an action to run when no results are produced.
-    runNonDetC :: forall b . (a -> m b -> m b) -> m b -> m b
+  { -- | A higher-order function receiving three continuations, respectively implementing '<|>', 'pure', and 'empty'.
+    runNonDetC :: forall b . (m b -> m b -> m b) -> (a -> m b) -> m b -> m b
   }
-  deriving stock (Functor)
+  deriving (Functor)
 
+-- $
+--   prop> run (runNonDet (pure a *> pure b)) === Just b
+--   prop> run (runNonDet (pure a <* pure b)) === Just a
 instance Applicative (NonDetC m) where
-  pure a = NonDetC (\ cons -> cons a)
+  pure a = NonDetC (\ _ leaf _ -> leaf a)
   {-# INLINE pure #-}
-  NonDetC f <*> NonDetC a = NonDetC $ \ cons ->
-    f (\ f' -> a (cons . f'))
+  NonDetC f <*> NonDetC a = NonDetC $ \ fork leaf nil ->
+    f fork (\ f' -> a fork (leaf . f') nil) nil
   {-# INLINE (<*>) #-}
 
+-- $
+--   prop> run (runNonDet (pure a <|> (pure b <|> pure c))) === Fork (Leaf a) (Fork (Leaf b) (Leaf c))
+--   prop> run (runNonDet ((pure a <|> pure b) <|> pure c)) === Fork (Fork (Leaf a) (Leaf b)) (Leaf c)
 instance Alternative (NonDetC m) where
-  empty = NonDetC (\ _ nil -> nil)
+  empty = NonDetC (\ _ _ nil -> nil)
   {-# INLINE empty #-}
-  NonDetC l <|> NonDetC r = NonDetC $ \ cons -> l cons . r cons
+  NonDetC l <|> NonDetC r = NonDetC $ \ fork leaf nil -> fork (l fork leaf nil) (r fork leaf nil)
   {-# INLINE (<|>) #-}
 
 instance Monad (NonDetC m) where
-  NonDetC a >>= f = NonDetC $ \ cons ->
-    a (\ a' -> runNonDetC (f a') cons)
+  NonDetC a >>= f = NonDetC $ \ fork leaf nil ->
+    a fork (\ a' -> runNonDetC (f a') fork leaf nil) nil
   {-# INLINE (>>=) #-}
 
 instance Fail.MonadFail m => Fail.MonadFail (NonDetC m) where
-  fail s = NonDetC (\ _ _ -> Fail.fail s)
+  fail s = lift (Fail.fail s)
   {-# INLINE fail #-}
 
 instance MonadFix m => MonadFix (NonDetC m) where
-  mfix f = NonDetC (\ cons nil -> mfix (\ a -> runNonDetC (f (head a)) (fmap . (:)) (pure [])) >>= foldr cons nil)
+  mfix f = NonDetC $ \ fork leaf nil ->
+    mfix (\ a -> runNonDetC (f (fromJust (fold (<|>) Just Nothing a)))
+      (liftA2 Fork)
+      (pure . Leaf)
+      (pure Nil))
+    >>= fold fork leaf nil
   {-# INLINE mfix #-}
 
 instance MonadIO m => MonadIO (NonDetC m) where
-  liftIO io = NonDetC (\ cons nil -> liftIO io >>= flip cons nil)
+  liftIO io = lift (liftIO io)
   {-# INLINE liftIO #-}
 
 instance MonadPlus (NonDetC m)
 
 instance MonadTrans NonDetC where
-  lift m = NonDetC (\ cons nil -> m >>= flip cons nil)
+  lift m = NonDetC (\ _ leaf _ -> m >>= leaf)
```

Comment:
We no longer invoke the `nil` continuation for `pure` & `lift`, which should make this slightly more efficient in general.